### PR TITLE
Add missing getter for virtual host (phase2)

### PR DESCRIPTION
**Pull request by echo304**
### Motivation:
https://github.com/line/armeria/issues/4722#issuecomment-1467701581
Apart from `multipartUploadsLocation`, there are still missing getters in `VirtualHost`, while `ServiceConfig` has them.
With this change, it probably can has more flexibility on setting properties.

### Modifications:
- Adds `defaultLogName` and `successFunction` properties to `VirtualHost`
- Change constructor signature of `VirtualHost`
- Fix codes where VirtualHost is instantiated
- Add test codes for `VirtualHostBuilder`

### Result:
- User will have more flexibility for setting of the properties

- As @minwoox mentioned, some test codes using reflection need to be written.

### CODE REVIEW

**Comment by echo304** [comment]
> To prevent this situation in the future, we can add a test that uses reflection to check if VirtualHost and ServiceConfig has the same property or not.
I was not able to figure out exactly how to write these kind of test code.
Can anyone give some examples or already existing test code from source code? 🙏

**Comment by minwoox** [comment]
> Can anyone give some examples or already existing test code from source code?

Sorry about the late reply. 😉 
We can use reflection to extract the method names from the `VirtualHost`:
```java
final Set<String> virtualHostMethods =
        ReflectionUtils.getMethods(VirtualHost.class, m -> Modifier.isPublic(m.getModifiers()))
                       .stream()
                       .map(Method::getName)
                       .collect(toImmutableSet());
```
We can use the same way for the `ServiceConfig` as well.
After that, we can compare both sets if they have all the methods that the counterpart has.
There are also exceptions that we need not compare such as `VirtualHost.port()`.

**Comment by echo304** [comment]
@minwoox Thanks for the reply.
And I found pretty similar test code from [here](https://github.com/echo304/armeria/blob/97aadce7e5e0cc4ad45681119a4332a8a8b5460d/core/src/test/java/com/linecorp/armeria/common/FlagsTest.java#L252)
Will write some test code soon

File: core/src/main/java/com/linecorp/armeria/server/VirtualHostBuilder.java [hunk 1]
```diff
@@ -1249,9 +1277,10 @@ VirtualHost build(VirtualHostBuilder template, DependencyInjector dependencyInje
         final VirtualHost virtualHost =
                 new VirtualHost(defaultHostname, hostnamePattern, port, sslContext(template),
                                 serviceConfigs, fallbackServiceConfig, rejectedRouteHandler,
-                                accessLoggerMapper, defaultServiceNaming, requestTimeoutMillis,
+                                accessLoggerMapper, defaultServiceNaming, defaultLogName, requestTimeoutMillis,
                                 maxRequestLength, verboseResponses, accessLogWriter,
-                                blockingTaskExecutor, multipartUploadsLocation, builder.build(),
+                                blockingTaskExecutor, successFunction, multipartUploadsLocation,
+                                builder.build(),
                                 requestIdGenerator);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/VirtualHostBuilder.java:1284]: Join two lines?

File: core/src/main/java/com/linecorp/armeria/server/VirtualHostBuilder.java [hunk 2]
```diff
@@ -955,6 +959,22 @@ public VirtualHostBuilder defaultServiceNaming(ServiceNaming defaultServiceNamin
...
+     *
+     * @param defaultLogName the default log name.
+     */
+    public VirtualHostBuilder defaultLogName(String defaultLogName) {
+        this.defaultLogName = requireNonNull(defaultLogName, "defaultLogName");
+        return this;
+    }
+
+    @VisibleForTesting
+    String defaultLogName() {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/VirtualHostBuilder.java:974]:
```suggestion
    @Nullable
    String defaultLogName() {
```

File: core/src/main/java/com/linecorp/armeria/server/VirtualHostBuilder.java [hunk 3]
```diff
@@ -1062,6 +1082,11 @@ public VirtualHostBuilder successFunction(SuccessFunction successFunction) {
         return this;
     }
 
+    @VisibleForTesting
+    SuccessFunction successFunction() {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/VirtualHostBuilder.java:1086]:
```suggestion
    @VisibleForTesting
    @Nullable
    SuccessFunction successFunction() {
```

File: core/src/test/java/com/linecorp/armeria/server/VirtualHostTest.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import org.junit.jupiter.api.Test;
+import org.reflections.ReflectionUtils;
+
+import com.google.common.collect.ImmutableSet;
+
+class VirtualHostTest {
+
+    @Test
+    void testApiConsistencyBetweenVirtualHostAndServiceConfig() {
+        //Check method consistency between VirtualHost and ServiceConfig excluding certain methods.
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/server/VirtualHostTest.java:35]:
```suggestion
        // Check method consistency between VirtualHost and ServiceConfig excluding certain methods.
```

File: core/src/test/java/com/linecorp/armeria/server/VirtualHostTest.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        final Set<String> virtualHostMethods =
+                ReflectionUtils.getMethods(VirtualHost.class, m -> Modifier.isPublic(m.getModifiers()))
+                               .stream()
+                               .map(Method::getName)
+                               .collect(Collectors.toSet());
+
+        final Set<String> knownIgnoreVirtualHostMethods = ImmutableSet.of(
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/server/VirtualHostTest.java:42]: `ignore` is a verb. How about simply?
```suggestion
        final Set<String> ignorableVirtualHostMethods = ImmutableSet.of(
```

File: core/src/main/java/com/linecorp/armeria/server/VirtualHost.java [hunk 6]
```diff
@@ -305,6 +314,21 @@ public ServiceNaming defaultServiceNaming() {
...
+    /**
+     * Returns the default value of the {@link RequestLog#name()} property which is used when no name was set
+     * via {@link RequestLogBuilder#name(String, String)}.
+     * If {@code null}, one of the following values will be used instead:
+     * <ul>
+     *   <li>gRPC - A capitalized method name defined in {@code io.grpc.MethodDescriptor}
+     *       (e.g, {@code GetItems})</li>
+     *   <li>Thrift and annotated service - a method name (e.g, {@code getItems})</li>
+     *   <li>{@link HttpService} - an HTTP method name</li>
+     * </ul>
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/VirtualHost.java:326]: This value never returns null so I think we can remove these lines.
```suggestion
```

File: core/src/test/java/com/linecorp/armeria/server/VirtualHostTest.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import java.lang.reflect.Modifier;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.junit.jupiter.api.Test;
+import org.reflections.ReflectionUtils;
+
+import com.google.common.collect.ImmutableSet;
+
+class VirtualHostTest {
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/server/VirtualHostTest.java:31]: How about renaming this to `VirtualHostAndServiceConfigConsistencyTest`?

**Review by minwoox — APPROVED** [review]
Left small suggestions. Thanks a lot, @echo304! 😄

File: core/src/test/java/com/linecorp/armeria/server/VirtualHostTest.java (hunk 7 above)
Comment (echo304) [reply core/src/test/java/com/linecorp/armeria/server/VirtualHostTest.java:31]: Sounds great!

**Review by ikhoon — COMMENTED** [review]
Thanks, @echo304!

**Review by ikhoon — APPROVED** [review]

**Comment by trustin** [comment]
@echo304  Oops! Sorry we introduced a conflict before merging your PR. Would you mind resolving the conflict before I merge this in? 🙇

**Review by trustin — APPROVED** [review]
Thanks a lot, @echo304! 🙇

**Review by jrhee17 — APPROVED** [review]
Thanks @echo304 👍 👍 👍

**Pull request merged**